### PR TITLE
THREE with CSS based auto resizing

### PR DESCRIPTION
**Pull request by greggman**
Here's what CSS based THREE might look like. I wish this was the default but making it the default would break stuff so you have to pass in `autoResize: true` when you make the `WebGLRenderer`
#6027

### CODE REVIEW

**Comment by greggman** [comment]
Does this make any sense? Any suggestions or comments? If you check out `webgl_geometry_colors.html` you'll see 8 lines (more if you count formatting) were removed including resize event handling. 

I thought I'd also provide one that actually needs resize as well, that's `webgl_geometry_text.html`. You can see it got simpler as well. 

Random things:

*   Should `autoResize` be an option to `WebGLRenderer` or go somewhere else
*   Should `EffectComposer` also have an `autoSize` option. Then calling `composer.reset()` would not be needed.
*   Should `PerspectiveCamera` check it its constructor if it has 3 or 4 arguments and if only 3 then assume they are `fov, zNear, zFar` (no aspect).

**Comment by WestLangley** [comment]
Sorry @greggman, but I thought you were going to show how to use the CSS-based sizing _and_ remove the pixel-ratio code from the renderer. (https://github.com/mrdoob/three.js/issues/6027#issuecomment-74178939).

**Comment by makc** [comment]
Grabbed my :fries: (while awaiting [popcorn](http://emojipedia.org/popcorn/))

**Comment by greggman** [comment]
Actually I said

> Of course that conflicts with the pixelRatio issue :(

I can't see how to put `autoSize` in THREE and rectify it with pixelRatio. I can see how to refactor all the samples so if pixelRatio is pulled out they use CSS for resizing. I suppose you can still auto-update the camera.

**Comment by WestLangley** [comment]
I am not sure the renderer should be modifying the camera properties. Also, an orthographic camera would need to be handled differently.

Would you be willing to file a PR to show an example of how you would implement the removal of pixel-ratio logic from the renderer, instead? I know you have been pushing for that.

**Comment by greggman** [comment]
Sure,

but FYI, orthographic cameras seem like they would require some kind of
units to be defined to decide how much to display. They take left, right,
top, bottom which are relative to nothing where as perspective cameras say
"make this the fov for a given canvas's height".

I suppose you could record the size of the canvas at the time the
orthographic camera's extents were set and then keep everything
proportional. It seems less likely to match the user's intent vs
perspective cameras though.

Also, you guys need to choose your goals. MrDoob seems to say they goal is
to do as much as possible for noobs so they don't have too.

So, it arguably easier to pass in 'autoSize' to either the Camera or the
Renderer and have the renderer update the camera at render time. Otherwise
there's no function in a camera that gets called automatically. You could
make one to be called manually like `camera.updateToMatchDisplay()` but
you've added a step for the user rather than removing one which if your
goal is as MrDoob says, make it simpler, then that's going away from that
goal not toward it.

**Comment by WestLangley** [comment]
> Also, you guys need to choose your goals. 

@mrdoob and I do not always agree. I like to use this forum to debate ideas, and then we let @mrdoob  have the final decision.

You have convinced me in prior posts that three.js should get out of the pixel-ratio-setting business. I would like to see how you would implement that in code, because I think your have good ideas. The only downside is @mrdoob will not like it, and you will have wasted some of your time.

**Comment by mrdoob** [comment]
I really like that autoResize is not `true` by default and that this maintains backwards compatibility. That way we can experiment with this and see if there are any downsides.

File: src/renderers/WebGLRenderer.js [hunk 1]
```diff
@@ -3310,6 +3341,21 @@ THREE.WebGLRenderer = function ( parameters ) {
...
+		if ( _sizeNeedsUpdate || _autoResize && ( renderTarget === undefined || renderTarget === null ) ) {
+
+			_sizeNeedsUpdate = false;
+
+			if ( this.resizeCanvasToMatchSizeItsDisplayed() ) {
+
+				camera.aspect = _canvas.clientWidth / _canvas.clientHeight;
+				camera.updateProjectionMatrix();
+
+				this.dispatchEvent( { type: 'resize' } );
```
Comment (mrdoob) [inline src/renderers/WebGLRenderer.js]: Shouldn't this be in `this.setSize`?

**Comment by threejsworker** [comment]
The examples of this pullrequest are now built and visible in threejsworker. To view them, go to the following link: 

 http://threejsworker.com/viewpullrequest.html#6073

**Comment by threejsworker** [comment]
The examples of this pullrequest are now built and visible in threejsworker. To view them, go to the following link: 

 http://threejsworker.com/viewpullrequest.html#6073

**Comment by threejsworker** [comment]
The examples of this pullrequest are now built and visible in threejsworker. To view them, go to the following link: 

 http://threejsworker.com/viewpullrequest.html#6073

**Comment by threejsworker** [comment]
The examples of this pullrequest are now built and visible in threejsworker. To view them, go to the following link: 

 http://threejsworker.com/viewpullrequest.html#6073

**Pull request closed without merging**